confection/__init__.py

```diff
@@ -889,30 +885,45 @@ def _fill(
             else:
                 filled[key] = value
                 # Prevent pydantic from consuming generator if part of a union
-                validation[v_key] = (
-                    value if not isinstance(value, GeneratorType) else []
-                )
+                validation[v_key] = value
                 final[key] = value
         # Now that we've filled in all of the promises, update with defaults
         # from schema, and validate if validation is enabled
         exclude = []
         if validate:
             try:
-                result = schema.parse_obj(validation)
+                result = schema.model_validate(validation)
             except ValidationError as e:
                 raise ConfigValidationError(
                     config=config, errors=e.errors(), parent=parent
                 ) from None
         else:
-            # Same as parse_obj, but without validation
-            result = schema.construct(**validation)
+            # Same as model_validate, but without validation
+            fields_set = set(schema.model_fields.keys())
+            result = schema.model_construct(fields_set, **validation)
             # If our schema doesn't allow extra values, we need to filter them
             # manually because .construct doesn't parse anything
-            if schema.Config.extra in (Extra.forbid, Extra.ignore):
-                fields = schema.__fields__.keys()
-                exclude = [k for k in result.__fields_set__ if k not in fields]
+            if schema.model_config.get("extra", "forbid") in ("forbid", "ignore"):
+                fields = result.model_fields_set
+                exclude = [k for k in dict(result).keys() if k not in fields]
         exclude_validation = set([ARGS_FIELD_ALIAS, *RESERVED_FIELDS.keys()])
-        validation.update(result.dict(exclude=exclude_validation))
+        # Do a shallow serialization first
+        # If any of the sub-objects are Pydantic models, first check if they
+        # were resolved earlier from a registry. If they weren't resolved
+        # they are part of a nested schema and need to be serialized with
+        # model.dict()
+        # Allows for returning Pydantic models from a registered function
+        shallow_result_dict = dict(result)
+        if result.model_extra is not None:
+            shallow_result_dict.update(result.model_extra)
+        result_dict = {}
+        for k, v in shallow_result_dict.items():
+            if k in exclude_validation:
+                continue
+            result_dict[k] = v
+            if isinstance(v, BaseModel) and k not in resolved_object_keys:
+                result_dict[k] = v.model_dump()
+        validation.update(result_dict)
```

Comment:
This change might seem a bit weird but it came up because Pydantic now treats Pydantic dataclasses and BaseModels the same way during serialization. 

In Pydantic v1, calling `model.dict()` with an instance of a `dataclass` would not JSON serialize that dataclass, it would just return that dataclass.

This allowed for our Optimizer README example to work and resolve that dataclass instance. 

e.g. with pydantic v1 this worked still:

```
import dataclasses
from typing import Union, Iterable
import catalogue
from confection import registry, Config

# Create a new registry.
registry.optimizers = catalogue.create("confection", "optimizers", entry_points=False)


# Define a dummy optimizer class.
@dataclasses.dataclass
class MyCoolOptimizer:
    learn_rate: float
    gamma: float


@registry.optimizers.register("my_cool_optimizer.v1")
def make_my_optimizer(learn_rate: Union[float, Iterable[float]], gamma: float):
    return MyCoolOptimizer(learn_rate, gamma)


# Load the config file from disk, resolve it and fetch the instantiated optimizer object.
config = Config().from_disk("./config.cfg")
resolved = registry.resolve(config)
optimizer = resolved["optimizer"]  # MyCoolOptimizer(learn_rate=0.001, gamma=1e-08)
```

However, when swapped to pydantic v2, the final line here was resolving the `optimzer` to a dict.

```python
...
optimizer = resolved["optimizer"]  # {"learn_rate": 0.001, "gamma": 1e-08}
```

This has actually always been a bug, because previously we could not make `MyCoolOptimizer` into a Pydantic model, it had to be a `dataclass` (or any other class that wasn't a Pydantic model). 

With this code above, we could make `MyCoolOptimzer` into a Pydantic model and we'd get a Pydantic model back.